libs/knowledge-store/ragstack_knowledge_store/graph_store.py

```diff
@@ -670,3 +751,113 @@ def add_targets(rows: Iterable[Any]) -> None:
+            fields = {_field_spec} if isinstance(_field_spec, str) else set(_field_spec)
+            if _mode.lower() in {
+                "default_to_unsearchable",
+                "allowlist",
+                "allow",
+                "allow_list",
+            }:
+                mode = MetadataIndexingMode.DEFAULT_TO_UNSEARCHABLE
+            elif _mode.lower() in {
+                "default_to_searchable",
+                "denylist",
+                "deny",
+                "deny_list",
+            }:
+                mode = MetadataIndexingMode.DEFAULT_TO_SEARCHABLE
+            else:
+                raise ValueError(
+                    f"Unsupported metadata indexing mode specification '{_mode}'"
+                )
+        return (mode, fields)
+
+    @staticmethod
+    def _coerce_string(value: Any) -> str:
+        if isinstance(value, str):
+            return value
+        if isinstance(value, bool):
+            # bool MUST come before int in this chain of ifs!
+            return json.dumps(value)
+        if isinstance(value, int):
+            # we don't want to store '1' and '1.0' differently
+            # for the sake of metadata-filtered retrieval:
+            return json.dumps(float(value))
+        if isinstance(value, float) or value is None:
+            return json.dumps(value)
+        # when all else fails ...
+        return str(value)
+
+    def _extract_where_clause_blocks(
+        self, metadata: Dict[str, Any]
+    ) -> Tuple[str, List[Any]]:
+        _, metadata_s = self._parse_metadata(metadata=metadata, is_query=True)
+
+        if len(metadata_s) == 0:
+            return "", []
+
+        wc_blocks: List[str] = []
+        vals_list: List[Any] = []
+
+        for k, v in sorted(metadata_s.items()):
+            wc_blocks.append(f"metadata_s['{k}'] = ?")
+            vals_list.append(v)
+
+        where_clause = "WHERE " + " AND ".join(wc_blocks)
+        return where_clause, vals_list
+
+    def _get_search_cql(
+        self,
+        limit: int,
+        columns: Optional[str] = CONTENT_COLUMNS,
+        metadata: Dict[str, Any] = {},
+        embedding: Optional[List[float]] = None,
+    ) -> Tuple[str, Tuple[Any, ...]]:
+        where_clause, get_cql_vals = self._extract_where_clause_blocks(
+            metadata=metadata
+        )
+        limit_clause = "LIMIT ?"
+        limit_cql_vals = [limit]
+
+        order_clause = ""
+        order_cql_vals = []
+        if embedding is not None:
+            order_clause = "ORDER BY text_embedding ANN OF ?"
+            order_cql_vals = [embedding]
+
+        select_vals = tuple(list(get_cql_vals) + order_cql_vals + limit_cql_vals)
+        select_cql = SELECT_CQL_TEMPLATE.format(
+            columns=columns,
+            table_name=self.table_name(),
+            where_clause=where_clause,
+            order_clause=order_clause,
+            limit_clause=limit_clause,
+        )
+        prepared_query = self._session.prepare(select_cql)
```

Comment:
We may need to cache these if the preparation takes too long. Specifically, this will re-preprae the same query multiple times.

Reply:
is it okay to save this for another PR?

Reply:
added to this PR.